L1Trigger/L1THGCal/interface/HGCalCoarseTriggerCellMapping.h

```diff
@@ -0,0 +1,70 @@
+#ifndef __L1Trigger_L1THGCal_HGCalCoarseTriggerCellMapping_h__
+#define __L1Trigger_L1THGCal_HGCalCoarseTriggerCellMapping_h__
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
+
+#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
+#include "DataFormats/L1THGCal/interface/HGCalTriggerSums.h"
+#include "DataFormats/ForwardDetId/interface/HGCSiliconDetIdToROC.h"
+
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerTools.h"
+      
+class HGCalCoarseTriggerCellMapping
+{
+  public:
+    HGCalCoarseTriggerCellMapping(const edm::ParameterSet& conf);
+    void setEvenDetId(l1t::HGCalTriggerCell &c) const;
+    uint32_t getEvenDetId(int tcid) const;
+    std::vector<uint32_t> getConstituentTriggerCellsByThickness( int ctcId, int thickness) const;
+    std::vector<uint32_t> getConstituentTriggerCells( int ctcId, int ctcSize) const;
+    void setCoarseTriggerCellPosition( l1t::HGCalTriggerCell& tc, const int coarse_size ) const;
+    int getCoarseTriggerCellId(int detid, int CTCsize = -1) const ;
+    int getCoarseTriggerCellIdByThickness(int detid, int thickness) const ;
```

Comment:
I think having `getCoarseTriggerCellId` without relying on the thickness could also be useful, as it gives more flexibility to someone in the future, who might want to divide based on another quantity, e.g. layer.
Regarding deriving the thickness from the detid - do I need to have run an `eventSetup` for the triggerTools member of this class? If so how is the best way to pass the `const edm::EventSetup& es`?

Reply:
If someone wants to divide according to an other quantity it will have to be implemented in this class, not outside. This is the goal of this class to encapsulate geometry and grouping issues such that the user of the class doesn't need to bother about that. Once this class is initialized, it should be able to return an id just using the input id.

Reply:
Concerning the thickness, you need a trigger tools object, initialized with an `eventSetup()` call as it is done elsewhere. In the end there are chains of `eventSetup()` calls from the processor class, to implementation classes down to trigger tools.

Reply:
Yes I do agree, but why couldn't we have both functions, and then later someone might add `getCoarseTriggerCellIdByLayer` which will also call `getCoarseTriggerCellId`.
The main issue I have is the line in `setCoarseTriggerCellPosition`:

 ```    std::vector<uint32_t> constituentTCs = getConstituentTriggerCells ( getCoarseTriggerCellId(tc.detId(), coarse_size ), coarse_size );```

Here coarse_size is not really related to the same STCsizes in the input vector

Reply:
> but why couldn't we have both functions, and then later someone might add getCoarseTriggerCellIdByLayer

- it complicates the interface (it's not clear which method the user should use) and it means also that the interface changes each time someone wants to modify the size dependencies.
- the user needs to write additional lines of code each time this method is called (e.g. to retrieve the thickness), while it could be written only once.
- the possibility to use a size which is different from the initialized one makes the code hard to follow.

> Here coarse_size is not really related to the same STCsizes in the input vector

Well, the size should always be the initialized size, otherwise it's confusing. The mapper class should be initialized with the correct sizes.

Reply:
I agree with the points you are making, I'm just not sure how to do it though. 
Previously when `stcSize_` was a function of the SuperTriggerCellImpl class the mapper was initialised with the correct sizes as you say.
The only thing I can think of is an optional parameter in the initialisation to overide the `stcSize_` when we want to use it for something else (e.g. coarse-2 cells).

Reply:
I see. I think it is just needed to change the constructor, taking a `vector<unsigned>` instead of `ParameterSet`.